### PR TITLE
[SYCL][FPGA] Add support for two new function attributes

### PR DESCRIPTION
**Pull request by smanna12**
This patch implements the support for two new FPGA function attributes:
  1. disable_loop_pipelining
  2. initiation_interval

Both attributes alreday exist for loops as statement attributes.

Function attribute "initiation_interval" takes a single unsigned integer argument.

Syntax:
[[intel::initiation_interval(n)]] void foo() {}

The argument can be constexpr. The parameter n is a template parameter.

The LLVM IR representation will be function metadata:
!initiation_interval !0
!0 = !{!i32 n}

Function attribute "disable_loop_pipelining" takes no arguments.

Syntax:
[[intel::disable_loop_pipelining]] void foo() {}

The LLVM IR representation will be function metadata:
!disable_loop_pipelining !0
!0 = !{!i32 1}

Both attributes are only valid in device code and do not get propagated to the caller.
The function attributes apply to a lambda function, or function definition.

Signed-off-by: Soumi Manna <smanna.manna@intel.com>

### CODE REVIEW

File: clang/include/clang/Basic/Attr.td [hunk 1]
```diff
@@ -1878,16 +1878,18 @@ def SYCLIntelFPGAIVDep : StmtAttr {
...
-  let Subjects = SubjectList<[ForStmt, CXXForRangeStmt, WhileStmt, DoStmt],
-                             ErrorDiag, "'for', 'while', and 'do' statements">;
+  let Subjects = SubjectList<[ForStmt, CXXForRangeStmt, WhileStmt, DoStmt, Function],
+                              ErrorDiag,
+                              "'for', 'while', 'do' statements, and functions">;
   let Args = [ExprArgument<"IntervalExpr", /*opt*/1>];
   let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
   let HasCustomTypeTransform = 1;
   let Documentation = [SYCLIntelFPGAInitiationIntervalAttrDocs];
+  let SupportsNonconformingLambdaSyntax = 1;
```
Comment (smanna12) [inline clang/include/clang/Basic/Attr.td:1892]: i checked with feature request author, the attribute needs to apply on lambda.

File: clang/include/clang/Basic/Attr.td [hunk 2]
```diff
@@ -1912,14 +1914,16 @@ def SYCLIntelFPGALoopCoalesce : StmtAttr {
...
                    CXX11<"intel","disable_loop_pipelining">];
-  let Subjects = SubjectList<[ForStmt, CXXForRangeStmt, WhileStmt, DoStmt],
-                             ErrorDiag, "'for', 'while', and 'do' statements">;
+  let Subjects = SubjectList<[ForStmt, CXXForRangeStmt, WhileStmt, DoStmt, Function],
+                              ErrorDiag,
+                              "'for', 'while', 'do' statements, and functions">;
   let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
   let HasCustomTypeTransform = 1;
   let Documentation = [SYCLIntelFPGADisableLoopPipeliningAttrDocs];
+  let SupportsNonconformingLambdaSyntax = 1;
```
Comment (smanna12) [inline clang/include/clang/Basic/Attr.td:1926]: same here too.

File: clang/include/clang/Basic/Attr.td (hunk 1 above)
Comment (AaronBallman) [reply clang/include/clang/Basic/Attr.td:1892]: Thanks for checking with the authors on these! Just to check -- when you ask their opinion, are you letting them know that a "yes" answer means their attribute isn't valid C++ code? (I have a suspicion that not everyone is thinking about the language design of these attributes so much as what immediate problem they think they're solving. Being asked if they want to add something that's not valid C++ may help them think a bit deeper about the problem.)

File: clang/include/clang/Basic/AttrDocs.td [hunk 3]
```diff
@@ -2832,9 +2832,9 @@ def SYCLIntelFPGAInitiationIntervalAttrDocs : Documentation {
   let Category = DocCatVariable;
   let Heading = "intel::initiation_interval";
   let Content = [{
-This attribute applies to a loop. Indicates that the loop should be pipelined
-with an initiation interval of N. N must be a positive integer. Cannot be
-applied multiple times to the same loop.
+This attribute applies to a loop or a function. Indicates that the loop or
+function should be pipelined with an initiation interval of N. N must be a
+positive integer. Cannot be applied multiple times to the same loop or function.
```
Comment (AaronBallman) [inline clang/include/clang/Basic/AttrDocs.td:2837]: Missing documentation about what other attributes this one is mutually exclusive with.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 4]
```diff
@@ -3330,6 +3330,105 @@ static void handleUseStallEnableClustersAttr(Sema &S, Decl *D,
...
 }
 
+// Handles disable_loop_pipelining attribute.
+static void handleSYCLIntelFPGADisableLoopPipeliningAttr(Sema &S, Decl *D,
+                                                         const ParsedAttr &A) {
+  S.CheckDeprecatedSYCLAttributeSpelling(A);
+
+  // [[intel::disable_loop_pipelining] and [[intel::initiation_interval()]]
+  // attributes are incompatible.
+  if (checkAttrMutualExclusion<SYCLIntelFPGAInitiationIntervalAttr>(S, D, A))
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3340]: The documentation says this attribute is incompatible with `max_interleaving, speculated_iterations,
max_concurrency, initiation_interval, or ivdep`, so I think there are some mutual exclusion checks missing.

Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:3340]: they are already available as loop attribute, we do not have any max_interleaving, speculated_iterations,or ivdep function attribute.

Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:3340]: Ah, thank you for pointing that out, that makes sense.

Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:3340]: I will update documentation.

File: clang/include/clang/Basic/AttrDocs.td (hunk 3 above)
Comment (smanna12) [reply clang/include/clang/Basic/AttrDocs.td:2837]: Done

**Comment by AaronBallman** [comment]
This generally LGTM but it's still marked as a draft -- are you planning more changes?

File: clang/test/SemaSYCL/initiation_interval.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+// Tests for Intel FPGA initiation_interval function attribute duplication.
+// No diagnostic is emitted because the arguments match. Duplicate attribute is silently ignored.
+[[intel::initiation_interval(2)]]
+[[intel::initiation_interval(2)]] void func3() {}
+
+// No diagnostic is emitted because the arguments match.
+[[intel::initiation_interval(4)]] void func4();
+[[intel::initiation_interval(4)]] void func4(); // OK
+
+// No diagnostic is emitted because the arguments match. Duplicate attribute is silently ignored.
```
Comment (AaronBallman) [inline clang/test/SemaSYCL/initiation_interval.cpp:23]: This comment is incorrect.

**Review by AaronBallman — APPROVED** [review]
LGTM aside from a commenting nit.

File: clang/test/SemaSYCL/initiation_interval.cpp (hunk 5 above)
Comment (smanna12) [reply clang/test/SemaSYCL/initiation_interval.cpp]: Done. Fixed now. Sorry about it.

File: clang/include/clang/Basic/AttrDocs.td [hunk 6]
```diff
@@ -2846,10 +2848,13 @@ The ``[[intel::ii]]`` attribute spelling is a deprecated synonym for
     [[intel::initiation_interval(4)]] for (int i = 0; i < 10; ++i) var++;
   }
 
+  [[intel::initiation_interval(4)]] void foo1 { }
+
   template<int N>
   void bar() {
     [[intel::initiation_interval(N)]] for(;;) { }
   }
+  [[intel::initiation_interval(N)]] void bar1 { }
```
Comment (premanandrao) [inline clang/include/clang/Basic/AttrDocs.td:2857]: Minor nit:  Please introduce the template argument N.

Comment (smanna12) [reply clang/include/clang/Basic/AttrDocs.td:2859]: Done. Thank you.

**Review by premanandrao — APPROVED** [review]
LGTM

**Comment by smanna12** [comment]
Thanks for the reviews.

**Pull request merged**